allennlp/training/trainer.py

```diff
@@ -126,6 +160,8 @@ def __init__(
         momentum_scheduler: Optional[MomentumScheduler] = None,
         tensorboard_writer: TensorboardWriter = None,
         moving_average: Optional[MovingAverage] = None,
+        batch_callback: Optional[BatchCallback] = None,
+        epoch_callback: Optional[EpochCallback] = None,
```

Comment:
You could make these `List`s instead of `Optional`, and call all of them. If we're going the callback route, we might as well go all in.

Reply:
Note that we are not providing any implementation of a callback here.  Anything a person wants to do will be code they write.  If they want to have several things done, they can just put those things in a callback.  If they want to have composable things in their callbacks, they have modularize it however they want (including a callback that calls a list of functions).

Reply:
I can imagine a few orthogonal, default callbacks we could provide. Can we implement `TensorboardWriter` as one of these? Writing predictions to a file? Automatic logging to W&B?

The advantage of using callbacks is configurability and composability. It would be a neat feature if you can just add what you need to this list of callbacks.

Reply:
Fair enough, you make a good point.  I don't think we can pull the tensorboard writer out into a callback, because of how computing gradient updates works, but you're right that you could imagine some built-in ones that get added later, and being able to pass in a list makes sense.  It was easy enough to change.